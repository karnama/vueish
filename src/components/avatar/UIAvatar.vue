<template>
    <div class="ui-avatar rounded-full text-lg overflow-hidden h-8 w-8
                text-white bg-red-500 flex flex-col justify-center items-center">
        <img v-show="src && loaded"
             :src="src"
             :alt="alt"
             @load="loaded = true"
             @error="loaded = false">
        <slot v-if="!loaded">
            <template v-if="alt">
                {{ alt.charAt(0).toUpperCase() }}
            </template>
            <template v-else>
                <span v-html="personIcon" />
            </template>
        </slot>
    </div>
</template>

<script lang="ts">
import { computed, defineComponent, ref } from 'vue';
import { getIcon } from '@/helpers';

export default defineComponent({
    name: 'UIAvatar',

    props: {
        /**
         * The alt text of the image.
         */
        alt: {
            type: String
        },

        /**
         * The source of the image.
         */
        src: {
            type: String
        }
    },

    setup(props) {
        const loaded = ref(false);
        const personIcon = getIcon('person');
        const showImg = computed(() => {
            return loaded.value && props.src;
        });

        return {
            loaded,
            personIcon,
            showImg
        };
    }
});
</script>
