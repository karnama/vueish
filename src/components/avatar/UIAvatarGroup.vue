<template>
    <slot v-if="false" />
    <div class="ui-avatar-group flex items-center -space-x-2" v-bind="omit($attrs, ['class'])">
        <template v-if="avatars.length">
            <UIAvatar v-for="(avatar, index) in avatars.slice(0, max)"
                      :key="index"
                      :src="avatar.src"
                      :alt="avatar.alt"
                      :squared="avatar.hasOwnProperty('squared') ? !!avatar.squared : squared"
                      :content="avatar.hasOwnProperty('content') ? String(avatar.content) : ''"
                      :class="$attrs.class"
                      :style="[!reverseStacking ? 'z-index: ' + (avatars.length - index) + '0;' : '']" />
        </template>
        <template v-else>
            <component :is="avatar"
                       v-for="(avatar, index) in normalisedAvatars.slice(0, max)"
                       :key="index"
                       :class="$attrs.class"
                       :style="[!reverseStacking ? 'z-index: ' + (normalisedAvatars.length - index) + '0;' : '']" />
        </template>
        <UIAvatar v-if="(avatars.length ? avatars.length : normalisedAvatars.length) > max"
                  :class="$attrs.class"
                  :content="'+' + String((avatars.length ? avatars.length : normalisedAvatars.length) - max)"
                  :style="[!reverseStacking ? 'z-index: 1' : '']">
            +{{ (avatars.length ? avatars.length : normalisedAvatars.length) - max }}
        </UIAvatar>
    </div>
</template>

<script lang="ts">
import { computed, defineComponent } from 'vue';
import type { PropType, VNode } from 'vue';
import UIAvatar from '@components/avatar/UIAvatar.vue';
import type { Avatar } from '@/types';
import { omit } from 'lodash-es';

export default defineComponent({
    name: 'UIAvatarGroup',

    components: { UIAvatar },

    inheritAttrs: false,

    props: {
        /**
         * The data for the avatars to create.
         */
        avatars: {
            type: Array as PropType<Avatar[]>,
            default: () => []
        },

        /**
         * The z index order of the avatars.
         */
        reverseStacking: {
            type: Boolean,
            default: false
        },

        /**
         * The maximum number of avatars to show.
         */
        max: {
            type: Number,
            default: 3
        },

        /**
         * Display the avatar with square styling.
         */
        squared: Boolean
    },

    setup(props, ctx) {
        const normalisedAvatars = computed<VNode[]>(() => {
            if (!ctx.slots.hasOwnProperty('default')) {
                return [];
            }

            return ctx.slots.default();
        });

        return {
            normalisedAvatars,
            omit
        };
    }
});
</script>

<style lang="scss">
.ui-avatar-group {
    & > .ui-avatar {
        border: 2px solid theme('colors.gray.100');
    }
}

.dark .ui-avatar-group {
    & > .ui-avatar {
        border: 2px solid theme('colors.gray.700');
    }
}
</style>
