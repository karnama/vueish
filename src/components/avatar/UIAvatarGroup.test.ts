import { mount } from '@vue/test-utils';
import UIAvatarGroup from '@components/avatar/UIAvatarGroup.vue';
import type { Avatar } from 'types';
import UIAvatar from '@components/avatar/UIAvatar.vue';

const avatars: Avatar[] = [
    {},
    { alt: 'name' },
    { content: 'something' },
    { src: 'imgSource' },
    { squared: true }
];

describe('UIAvatarGroup', () => {
    it('should correctly display', () => {
        const wrapper = mount(UIAvatarGroup, {
            props: {
                avatars,
                max: avatars.length
            }
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should display the avatars given as prop', () => {
        const wrapper = mount(UIAvatarGroup, {
            props: {
                avatars,
                max: avatars.length
            }
        });

        expect(wrapper.findAll('.ui-avatar')).toHaveLength(avatars.length);
    });

    // currently not possible to test: https://github.com/vuejs/vue-test-utils-next/issues/482
    // eslint-disable-next-line jest/no-disabled-tests
    it.skip('should display the avatars given in the default slot', () => {
        const wrapper = mount(UIAvatarGroup, {
            props: {
                max: avatars.length
            },
            slots: {
                default: [UIAvatar]
            }
        });

        expect(wrapper.findAll('.ui-avatar')).toHaveLength(avatars.length);
    });

    it('should apply classes to the avatars', () => {
        const wrapper = mount(UIAvatarGroup, {
            props: {
                avatars,
                max: avatars.length
            },
            attrs: {
                class: 'my-custom-class'
            }
        });

        const mountedAvatars = wrapper.findAll('.ui-avatar')
            .filter(avatar => avatar.html().includes('my-custom-class'));
        expect(mountedAvatars).toHaveLength(avatars.length);
    });

    it('should display max number of avatars as specified by the prop', async () => {
        const wrapper = mount(UIAvatarGroup, {
            props: {
                avatars
            }
        });

        expect(wrapper.findAll('.ui-avatar')).toHaveLength(4); // default 3 + 1
        await wrapper.setProps({ max: 1 });
        expect(wrapper.findAll('.ui-avatar')).toHaveLength(2);
        await wrapper.setProps({ max: avatars.length });
        expect(wrapper.findAll('.ui-avatar')).toHaveLength(avatars.length);
    });

    it('should set the rounded prop on the avatars', () => {
        const wrapper = mount(UIAvatarGroup, {
            props: {
                avatars,
                rounded: true
            }
        });

        expect(
            wrapper.findAll('.ui-avatar').every(avatar => avatar.html().includes('rounded-full'))
        ).toBe(true);
    });
});
