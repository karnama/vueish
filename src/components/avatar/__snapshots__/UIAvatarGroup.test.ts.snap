// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIAvatarGroup should correctly display 1`] = `
<div
  data-v-app=""
>
  
  <!--v-if-->
  <div
    class="ui-avatar-group flex items-center -space-x-2"
  >
    
    <div
      class="ui-avatar text-lg overflow-hidden h-8 w-8 text-white bg-brand-400 flex flex-col justify-center items-center rounded-full"
      style="z-index: 50;"
    >
      
      <span>
        <svg
          aria-label="person icon"
          fill="currentColor"
          height="18px"
          viewBox="0 0 24 24"
          width="18px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v1c0 .55.45 1 1 1h14c.55 0 1-.45 1-1v-1c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </span>
      
    </div>
    <div
      class="ui-avatar text-lg overflow-hidden h-8 w-8 text-white bg-brand-400 flex flex-col justify-center items-center rounded-full"
      style="z-index: 40;"
    >
      
      
      N
      
      
    </div>
    <div
      class="ui-avatar text-lg overflow-hidden h-8 w-8 text-white bg-brand-400 flex flex-col justify-center items-center rounded-full"
      style="z-index: 30;"
    >
      
      
      something
      
      
    </div>
    <div
      class="ui-avatar text-lg overflow-hidden h-8 w-8 text-white bg-brand-400 flex flex-col justify-center items-center rounded-full"
      style="z-index: 20;"
    >
      <img
        src="imgSource"
      />
    </div>
    <div
      class="ui-avatar text-lg overflow-hidden h-8 w-8 text-white bg-brand-400 flex flex-col justify-center items-center rounded"
      style="z-index: 10;"
    >
      
      <span>
        <svg
          aria-label="person icon"
          fill="currentColor"
          height="18px"
          viewBox="0 0 24 24"
          width="18px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v1c0 .55.45 1 1 1h14c.55 0 1-.45 1-1v-1c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </span>
      
    </div>
    
    <!--v-if-->
  </div>
  
</div>
`;
